File: client/lib/models/structure.dart
import 'package:json_annotation/json_annotation.dart';
import '../game/hex/cube_coord.dart';
import 'enums.dart';

part 'structure.g.dart';

@JsonSerializable()
class Structure {
  final String id;
  final StructureType type;
  @JsonKey(name: 'owner_id')
  final String ownerId;
  @JsonKey(fromJson: CubeCoord.fromJson)
  final CubeCoord hex;
  @JsonKey(name: 'current_hp')
  final int currentHp;
  @JsonKey(name: 'max_hp')
  final int maxHp;
  final int atk;
  final int def;
  final int range;
  final String damage;
  final int income;
  @JsonKey(name: 'can_spawn')
  final bool canSpawn;

  const Structure({
    required this.id,
    required this.type,
    required this.ownerId,
    required this.hex,
    required this.currentHp,
    required this.maxHp,
    required this.atk,
    required this.def,
    required this.range,
    required this.damage,
    required this.income,
    required this.canSpawn,
  });

  bool get isNeutral => ownerId.isEmpty;
  bool isOwnedBy(String playerId) => ownerId == playerId;
  bool get isAlive => currentHp > 0;

  factory Structure.fromJson(Map<String, dynamic> json) =>
      _$StructureFromJson(json);
  Map<String, dynamic> toJson() => _$StructureToJson(this);

  Structure copyWith({
    String? id,
    StructureType? type,
    String? ownerId,
    CubeCoord? hex,
    int? currentHp,
    int? maxHp,
    int? atk,
    int? def,
    int? range,
    String? damage,
    int? income,
    bool? canSpawn,
  }) {
    return Structure(
      id: id ?? this.id,
      type: type ?? this.type,
      ownerId: ownerId ?? this.ownerId,
      hex: hex ?? this.hex,
      currentHp: currentHp ?? this.currentHp,
      maxHp: maxHp ?? this.maxHp,
      atk: atk ?? this.atk,
      def: def ?? this.def,
      range: range ?? this.range,
      damage: damage ?? this.damage,
      income: income ?? this.income,
      canSpawn: canSpawn ?? this.canSpawn,
    );
  }
}

File: server/internal/model/structure.go
package model

import (
	"github.com/teomiscia/hexbattle/internal/hex"
)

// Structure represents a capturable building on the map.
type Structure struct {
	ID        string        `json:"id"`
	Type      StructureType `json:"type"`
	OwnerID   string        `json:"owner_id"` // Empty string = neutral
	Hex       hex.Coord     `json:"hex"`
	CurrentHP int           `json:"current_hp"`
	MaxHP     int           `json:"max_hp"`
	ATK       int           `json:"atk"`
	DEF       int           `json:"def"`
	Range     int           `json:"range"`
	Damage    string        `json:"damage"` // Dice notation
	Income    int           `json:"income"`
	CanSpawn  bool          `json:"can_spawn"`
}

// IsNeutral returns true if the structure has no owner.
func (s *Structure) IsNeutral() bool {
	return s.OwnerID == ""
}

// IsOwnedBy returns true if the structure is owned by the given player.
func (s *Structure) IsOwnedBy(playerID string) bool {
	return s.OwnerID == playerID
}

// IsAlive returns true if the structure has HP remaining.
func (s *Structure) IsAlive() bool {
	return s.CurrentHP > 0
}

// TakeDamage applies damage to the structure and returns true if it reached 0 HP.
func (s *Structure) TakeDamage(amount int) bool {
	s.CurrentHP -= amount
	if s.CurrentHP < 0 {
		s.CurrentHP = 0
	}
	return s.CurrentHP <= 0
}

// Capture transfers ownership to the given player and sets HP to max.
func (s *Structure) Capture(newOwnerID string) {
	s.OwnerID = newOwnerID
	s.CurrentHP = s.MaxHP
}

// Heal restores HP up to MaxHP. Returns the amount actually healed.
func (s *Structure) Heal(amount int) int {
	before := s.CurrentHP
	s.CurrentHP += amount
	if s.CurrentHP > s.MaxHP {
		s.CurrentHP = s.MaxHP
	}
	return s.CurrentHP - before
}
